File: docs/dfs/2876_hard/visited_nodes.py

class DirectedGraphVisitedNodes:  # LeetCode Q.2876.
    def __init__(self):
        self.graph: list[int] = []
        self.visited_orders: list[int] = []
        self.current_order = 1
        self.visited_counts: list[int] = []

    def _dfs_visited_nodes(self, node: int) -> int:
        self.visited_orders[node] = self.current_order
        self.current_order += 1

        cycle_start_order = -1

        target_node = self.graph[node]
        if self.visited_orders[target_node] == -1:  # Unvisited target node.
            cycle_start_order = self._dfs_visited_nodes(target_node)

        # A cycle just forms. Current node and its target node are members.
        if self.visited_orders[target_node] > 0:
            cycle_size = self.visited_orders[node] + 1 - self.visited_orders[target_node]
            self.visited_counts[node] = cycle_size

            self.visited_orders[node] = -2  # Current node has finalized DFS.

            return self.visited_orders[target_node]  # Where cycle starts.

        # Reset: current node is no part of any cycle. Nor are its ascendants.
        if self.visited_orders[node] < cycle_start_order:
            cycle_start_order = -1

        self.visited_counts[node] = self.visited_counts[target_node]
        if cycle_start_order == -1:
            self.visited_counts[node] += 1

        self.visited_orders[node] = -2  # Current node has finalized DFS.
        return cycle_start_order

    def count_visited_nodes(self, edges: list[int]) -> list[int]:
        self.graph = edges
        self.visited_counts = [0] * len(edges)

        # Special marks: -1 = unvisited yet. -2 = has finalized DFS.
        self.visited_orders = [-1] * len(edges)  # -1 means unvisited yet.
        self.current_order = 1

        for node in range(len(edges)):
            if self.visited_orders[node] == -1:
                self._dfs_visited_nodes(node)

        return self.visited_counts

File: docs/dfs/2876_hard/visited_nodes.cpp
#include <vector>
using namespace std;

class DirectedGraphVisitedNodes // LeetCode Q.2876.
{
private:
    vector<int> graph, visitedOrders, visitedCounts;
    int currentOrder = 1;

    int dfsVisitedNodes(int node) {
        visitedOrders[node] = currentOrder;
        currentOrder++;

        int cycleStartOrder = -1;

        int tgtNode = graph[node];
        if (visitedOrders[tgtNode] == -1) // Unvisited target node.
            cycleStartOrder = dfsVisitedNodes(tgtNode);

        // A cycle just forms. Current node and its target node are members.
        if (visitedOrders[tgtNode] > 0) {
            int cycleSize = visitedOrders[node] + 1 - visitedOrders[tgtNode];
            visitedCounts[node] = cycleSize;

            visitedOrders[node] = -2; // Current node has finalized DFS.

            return visitedOrders[tgtNode]; // Where cycle starts.
        }

        // Reset: current node is no part of any cycle. Nor are its ascendants.
        if (visitedOrders[node] < cycleStartOrder)
            cycleStartOrder = -1;

        visitedCounts[node] = visitedCounts[tgtNode];
        if (cycleStartOrder == -1)
            visitedCounts[node]++;

        visitedOrders[node] = -2; // Current node has finalized DFS.
        return cycleStartOrder;
    }

public:
    vector<int> countVisitedNodes(vector<int>& edges) {
        graph = edges;
        visitedCounts.assign(edges.size(), 0);

        // Special marks: -1 = unvisited yet. -2 = has finalized DFS.
        visitedOrders.assign(edges.size(), -1);

        for (int node = 0; node < graph.size(); node++)
            if (visitedOrders[node] == -1)
                dfsVisitedNodes(node);

        return visitedCounts;
    }
};